stablehlo/dialect/TypeInference.cpp

```diff
@@ -1277,7 +1278,34 @@ static LogicalResult inferGatherReturnTypeComponents(
                             offsetDims, collapsedSliceDims, startIndexMap,
                             indexVectorDim, shape);
 
-  inferredReturnShapes.emplace_back(shape, elementType);
+  // The dimension sizes of result, corresponding to offset dimensions, depend
+  // on attributes (like `collapsed_slice_dims` and `slice_sizes`) and hence are
+  // always static. Whereas, the dimension sizes of result, corresponding to
+  // batch dimensions, depends on input `start_indices` and could be dynamic.
+  // The corresponding bounds, in that case,  are propagated from the
+  // `start_indices`.
+  ArrayRef<int64_t> startIndicesBounds = encodingToBounds(
+      startIndices.getType().cast<RankedTensorType>().getEncoding());
```

Comment:
Would be good to extract `startIndices.getType().cast<RankedTensorType>().getEncoding()` into a temporary variable like `startIndicesEncoding` since it's a pretty sizable expression and is used more than once.